File: v4/rust/iprange-cli/src/rpc/schema.rs
//! Strict v1 envelope decoding, response encoding, and error payloads.
//!
//! Everything here is transport contract (iprange-jsonrpc-v1.md):
//! - requests: `jsonrpc:"2.0"`, string or integral id, method prefix
//!   `iprange.v1.`, object params, unknown members rejected;
//! - notifications: only `iprange.v1.cancel` may omit the id;
//! - batches: 1..=16 objects, executed in array order, valid
//!   notifications excluded from the response array;
//! - responses: id required, exactly one of result/error, error
//!   objects carry integer code + string message + optional data;
//! - ceilings: response object 65,000 bytes; response frame
//!   1,048,576 bytes.

use serde_json::{json, Value};

use super::framing::{OUTPUT_FRAME_LIMIT, RESPONSE_OBJECT_LIMIT};

pub const STD_PARSE_ERROR: i64 = -32700;
pub const STD_INVALID_REQUEST: i64 = -32600;
pub const STD_METHOD_NOT_FOUND: i64 = -32601;
pub const STD_INVALID_PARAMS: i64 = -32602;
pub const TRANSPORT_FRAME_TOO_LARGE: i64 = -32001;
pub const TRANSPORT_SERVER_BUSY: i64 = -32002;
pub const PRODUCT_ERROR: i64 = -32010;

pub const METHOD_PREFIX: &str = "iprange.v1.";
pub const CANCEL_METHOD: &str = "iprange.v1.cancel";

/// A request id: an exact JSON string or an integral JSON number.
///
/// serde_json preserves i64 and u64 exactly, which covers every integral
/// JSON number the v1 contract accepts (the Python authority parses any
/// Python int; JSON text cannot express values beyond u64 with exact
/// integral semantics here).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RequestId {
    String(String),
    Number(serde_json::Number),
}

impl Default for RequestId {
    /// Only used to satisfy derive(Default) on connection state; the
    /// active id is always overwritten before a handler runs.
    fn default() -> Self {
        RequestId::String(String::new())
    }
}

impl RequestId {
    pub fn as_json(&self) -> Value {
        match self {
            RequestId::String(s) => Value::String(s.clone()),
            RequestId::Number(n) => Value::Number(n.clone()),
        }
    }
    /// Canonical correlation key used for cancellation and busy tracking.
    pub fn key(&self) -> String {
        match self {
            RequestId::String(s) => format!("s:{s}"),
            RequestId::Number(n) => format!("n:{n}"),
        }
    }
}

/// A decoded and envelope-valid request or notification.
#[derive(Debug, Clone)]
pub struct Request {
    pub id: Option<RequestId>,
    pub method: String,
    pub params: Value,
    pub batch_index: Option<usize>,
}

#[derive(Debug)]
pub struct SchemaError {
    pub code: i64,
    pub message: String,
}

impl SchemaError {
    pub fn parse(message: impl Into<String>) -> Self {
        Self {
            code: STD_PARSE_ERROR,
            message: message.into(),
        }
    }
    pub fn invalid(message: impl Into<String>) -> Self {
        Self {
            code: STD_INVALID_REQUEST,
            message: message.into(),
        }
    }
    pub fn response(id: Option<RequestId>, err: SchemaError) -> Value {
        let mut payload =
            json!({"jsonrpc": "2.0", "error": {"code": err.code, "message": err.message}});
        match &id {
            Some(id) => payload["id"] = id.as_json(),
            None => payload["id"] = Value::Null,
        }
        payload
    }
}

/// True when the preserved number text is an integral literal
/// (arbitrary_precision stores the exact lexical token).
pub(crate) fn integral_text(n: &serde_json::Number) -> bool {
    let text = n.to_string();
    !text.contains('.') && !text.contains('e') && !text.contains('E')
}

/// The cancellation/queue key of a `cancel.request_id` number: the same
/// canonical correlation key a request id of the same numeric text would
/// produce, so arbitrary-precision ids cancel correctly.
pub(crate) fn number_cancel_key(n: &Value) -> Option<String> {
    let number = n.as_number()?;
    if number.is_i64() || number.is_u64() || integral_text(number) {
        Some(format!("n:{number}"))
    } else {
        None
    }
}

fn valid_id(v: &Value) -> Option<RequestId> {
    match v {
        Value::String(s) => Some(RequestId::String(s.clone())),
        // Accept any integral JSON number. serde_json runs with
        // arbitrary_precision in this crate, so Number preserves the
        // exact decimal text of out-of-range integers; only true
        // floats (containing '.', 'e', or 'E') are rejected.
        Value::Number(n) if n.is_i64() || n.is_u64() || integral_text(n) => {
            Some(RequestId::Number(n.clone()))
        }
        _ => None,
    }
}

fn decode_envelope(item: &Value, batch_index: Option<usize>) -> Result<Request, SchemaError> {
    let obj = item
        .as_object()
        .ok_or_else(|| SchemaError::invalid("request must be an object"))?;
    for key in obj.keys() {
        if key != "jsonrpc" && key != "id" && key != "method" && key != "params" {
            return Err(SchemaError::invalid(format!(
                "unknown request member {key:?}"
            )));
        }
    }
    if obj.get("jsonrpc").and_then(|v| v.as_str()) != Some("2.0") {
        return Err(SchemaError::invalid("jsonrpc must be \"2.0\""));
    }
    let method = match obj.get("method") {
        Some(Value::String(m)) if m.starts_with(METHOD_PREFIX) => m.clone(),
        Some(Value::String(_)) => {
            return Err(SchemaError {
                code: STD_METHOD_NOT_FOUND,
                message: "method must start with iprange.v1.".into(),
            });
        }
        _ => {
            return Err(SchemaError::invalid(
                "method must be a string starting with iprange.v1.",
            ))
        }
    };
    let id = match obj.get("id") {
        None => None,
        Some(v) => Some(
            valid_id(v)
                .ok_or_else(|| SchemaError::invalid("id must be a string or integral number"))?,
        ),
    };
    let params = match obj.get("params") {
        Some(Value::Object(_)) => obj.get("params").unwrap().clone(),
        _ => {
            return Err(SchemaError {
                code: STD_INVALID_PARAMS,
                message: "params must be an object".into(),
            })
        }
    };
    if id.is_none() && method != CANCEL_METHOD {
        return Err(SchemaError::invalid(
            "notifications are not accepted except iprange.v1.cancel",
        ));
    }
    Ok(Request {
        id,
        method,
        params,
        batch_index,
    })
}

/// Decode one physical line into a single request or an ordered batch.
pub fn decode_frame(line: &[u8]) -> Result<Vec<Request>, SchemaError> {
    if line.len() > super::framing::INPUT_FRAME_LIMIT {
        return Err(SchemaError {
            code: TRANSPORT_FRAME_TOO_LARGE,
            message: "frame over input limit".into(),
        });
    }
    // Only LF and CRLF terminate a frame; a bare trailing CR is part
    // of the payload and is rejected below as an unescaped line break.
    let line = match line.strip_suffix(b"\n") {
        Some(rest) => match rest.strip_suffix(b"\r") {
            Some(crlf) => crlf,
            None => rest,
        },
        None => line,
    };
    if line.contains(&b'\n') || line.contains(&b'\r') {
        return Err(SchemaError::parse("unescaped line break inside frame"));
    }
    // The transport is a UTF-8 byte stream; reject invalid UTF-8 with
    // a deterministic parse error instead of letting the JSON decoder
    // apply a lossy replacement.
    std::str::from_utf8(line).map_err(|_| SchemaError::parse("frame is not valid UTF-8"))?;
    let value: Value = serde_json::from_slice(line)
        .map_err(|e| SchemaError::parse(format!("parse error: {e}")))?;
    if let Some(arr) = value.as_array() {
        if arr.is_empty() || arr.len() > super::framing::BATCH_LIMIT {
            return Err(SchemaError::invalid(format!(
                "batch length {} outside 1..{}",
                arr.len(),
                super::framing::BATCH_LIMIT
            )));
        }
        return arr
            .iter()
            .enumerate()
            .map(|(i, item)| decode_envelope(item, Some(i)))
            .collect();
    }
    if !value.is_object() {
        return Err(SchemaError::invalid("request must be an object"));
    }
    Ok(vec![decode_envelope(&value, None)?])
}

/// Encode one response object with the 65,000-byte ceiling.
pub fn encode_response_object(payload: &Value) -> Result<String, SchemaError> {
    let text = payload.to_string();
    if text.len() > RESPONSE_OBJECT_LIMIT {
        return Err(SchemaError {
            code: TRANSPORT_FRAME_TOO_LARGE,
            message: "response object over 65,000-byte limit".into(),
        });
    }
    Ok(text)
}

/// Encode one response frame with the 1,048,576-byte ceiling.
pub fn encode_response_frame(payload: &Value) -> Result<String, SchemaError> {
    let text = payload.to_string();
    if text.len() > OUTPUT_FRAME_LIMIT {
        return Err(SchemaError {
            code: TRANSPORT_FRAME_TOO_LARGE,
            message: "response frame over 1,048,576-byte limit".into(),
        });
    }
    Ok(text)
}

pub fn success_response(id: &RequestId, result: Value) -> Value {
    json!({"jsonrpc": "2.0", "id": id.as_json(), "result": result})
}

pub fn error_response(id: &RequestId, code: i64, message: &str, data: Option<Value>) -> Value {
    let mut payload = json!({"jsonrpc": "2.0", "id": id.as_json(),
                             "error": {"code": code, "message": message}});
    if let Some(data) = data {
        payload["error"]["data"] = data;
    }
    payload
}

#[cfg(test)]
mod tests {
    use super::*;

    fn decode(s: &str) -> Result<Vec<Request>, SchemaError> {
        decode_frame(s.as_bytes())
    }

    #[test]
    fn accepts_minimal_requests() {
        let reqs = decode(
            r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe","params":{}}"#,
        )
        .unwrap();
        assert_eq!(reqs.len(), 1);
        assert_eq!(reqs[0].method, "iprange.v1.system.describe");
    }

    #[test]
    fn rejects_envelope_violations() {
        for bad in [
            r#"{"id":"1","method":"iprange.v1.system.describe","params":{}}"#,
            r#"{"jsonrpc":"1.0","id":"1","method":"iprange.v1.system.describe","params":{}}"#,
            r#"{"jsonrpc":"2.0","id":1.5,"method":"iprange.v1.system.describe","params":{}}"#,
            r#"{"jsonrpc":"2.0","id":true,"method":"iprange.v1.system.describe","params":{}}"#,
            r#"{"jsonrpc":"2.0","id":"1","method":"other.x","params":{}}"#,
            r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe","params":[],"extra":1}"#,
            r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe","params":null}"#,
            r#"{"jsonrpc":"2.0","id":null,"method":"iprange.v1.system.describe","params":{}}"#,
        ] {
            assert!(decode(bad).is_err(), "accepted {bad}");
        }
    }

    #[test]
    fn params_are_required() {
        // The v1 contract says every request has object-valued params.
        assert!(
            decode(r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe"}"#).is_err()
        );
        assert!(decode(r#"{"jsonrpc":"2.0","method":"iprange.v1.cancel"}"#).is_err());
    }

    #[test]
    fn notification_only_cancel() {
        assert!(
            decode(r#"{"jsonrpc":"2.0","method":"iprange.v1.system.describe","params":{}}"#)
                .is_err()
        );
        let reqs =
            decode(r#"{"jsonrpc":"2.0","method":"iprange.v1.cancel","params":{"request_id":"1"}}"#)
                .unwrap();
        assert!(reqs[0].id.is_none());
        assert_eq!(reqs[0].method, CANCEL_METHOD);
    }

    #[test]
    fn invalid_utf8_is_a_deterministic_parse_error() {
        let err = decode_frame(b"{\"id\":\"\xff\xfe\"}").unwrap_err();
        assert_eq!(err.code, STD_PARSE_ERROR);
        assert_eq!(err.message, "frame is not valid UTF-8");
        let err = decode_frame(b"\xff").unwrap_err();
        assert_eq!(err.code, STD_PARSE_ERROR);
        assert_eq!(err.message, "frame is not valid UTF-8");
    }

    #[test]
    fn crlf_terminated_frames_decode() {
        let inner =
            r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe","params":{}}"#;
        let reqs = decode_frame(format!("{inner}\r\n").as_bytes()).unwrap();
        assert_eq!(reqs.len(), 1);
        assert_eq!(reqs[0].method, "iprange.v1.system.describe");
    }

    #[test]
    fn batch_bounds() {
        let inner =
            r#"{"jsonrpc":"2.0","id":"1","method":"iprange.v1.system.describe","params":{}}"#;
        let reqs = decode(&format!("[{inner},{inner}]")).unwrap();
        assert_eq!(reqs.len(), 2);
        // A single-element batch is valid (1..=16).
        assert_eq!(decode(&format!("[{inner}]")).unwrap().len(), 1);
        assert!(decode("[]").is_err());
    }
}

File: v4/go/internal/cli/rpc/schema.go
// Strict v1 envelope decoding, response encoding, and error payloads
// (iprange-jsonrpc-v1.md): requests carry `jsonrpc:"2.0"`, a string
// or integral id, a method with the `iprange.v1.` prefix, and
// object params; unknown members are rejected. Only the cancel
// notification may omit the id. Batches hold 1..=16 objects and
// execute in array order with valid notifications excluded from the
// response array. Responses require an id and exactly one of
// result/error; error objects carry integer code plus string message
// plus optional data. Ceilings: 65,000 bytes per response object and
// 1,048,576 bytes per response frame.

package rpc

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"unicode/utf8"
)

const (
	StdParseError          int64 = -32700
	StdInvalidRequest      int64 = -32600
	StdMethodNotFound      int64 = -32601
	StdInvalidParams       int64 = -32602
	TransportFrameTooLarge int64 = -32001
	TransportServerBusy    int64 = -32002
	ProductError           int64 = -32010

	MethodPrefix = "iprange.v1."
	CancelMethod = "iprange.v1.cancel"
)

// maxRequestDiagnosticBytes caps request-derived text echoed in an
// error message: the transport never echoes unbounded request bytes
// into a response (an error object must stay inside the response
// object ceiling).
const maxRequestDiagnosticBytes = 4096

// boundedDiagnosticText truncates request-derived diagnostic text with
// an explicit marker so an error object can never reach the response
// ceiling by echoing request bytes.
func boundedDiagnosticText(text string) string {
	if len(text) <= maxRequestDiagnosticBytes {
		return text
	}
	return text[:maxRequestDiagnosticBytes] + "...(truncated)"
}

// RequestId is an exact JSON string or an integral JSON number. The
// raw text is preserved so every integral literal the client sends is
// echoed byte-for-byte, except the literal -0, which is echoed as 0 to
// match the Rust serde_json normalization (both the response echo and
// the cancellation correlation key use the same canonical form).
type RequestId struct {
	// IsString selects the string form; otherwise the id is a number.
	IsString bool
	// Text is the exact JSON text (without quotes for strings).
	Text string
}

func RequestIdFromString(s string) RequestId { return RequestId{IsString: true, Text: s} }
func RequestIdFromNumber(n string) RequestId { return RequestId{IsString: false, Text: n} }

// AsJSON renders the id as a JSON value.
func (id RequestId) AsJSON() json.RawMessage {
	if id.IsString {
		b, _ := Marshal(id.Text)
		return b
	}
	return json.RawMessage(id.Text)
}

// Key is the canonical correlation key for cancellation and busy
// tracking: string ids and integral ids with the same text never
// collide.
func (id RequestId) Key() string {
	if id.IsString {
		return "s:" + id.Text
	}
	return "n:" + id.Text
}

// Request is one decoded, envelope-valid request or notification.
type Request struct {
	ID     *RequestId // nil for notifications
	Method string
	Params json.RawMessage
	// BatchIndex is set for members of a batch array.
	BatchIndex *int
}

// SchemaError is an envelope-level failure with a stable JSON-RPC
// code. Parse errors are -32700, invalid requests -32600.
type SchemaError struct {
	Code    int64
	Message string
}

func ParseError(message string) *SchemaError {
	return &SchemaError{Code: StdParseError, Message: message}
}
func InvalidRequest(message string) *SchemaError {
	return &SchemaError{Code: StdInvalidRequest, Message: message}
}

// Response builds the error envelope for id (or id null).
func (err *SchemaError) Response(id *RequestId) json.RawMessage {
	m := map[string]any{
		"jsonrpc": "2.0",
		"error":   map[string]any{"code": err.Code, "message": err.Message},
	}
	if id == nil {
		m["id"] = nil
	} else {
		m["id"] = json.RawMessage(id.AsJSON())
	}
	return mustMarshal(m)
}

// integralText reports whether the JSON number raw text is an integer
// literal (no '.', 'e', or 'E').
func integralText(raw json.RawMessage) bool {
	t := bytes.TrimSpace(raw)
	if len(t) == 0 {
		return false
	}
	for _, c := range t {
		switch c {
		case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '-', '+':
		default:
			return false
		}
	}
	return true
}

// isObjectRaw reports whether the raw JSON value is an object.
func isObjectRaw(raw json.RawMessage) bool {
	t := bytes.TrimSpace(raw)
	return len(t) > 0 && t[0] == '{'
}

// isStringRaw reports whether the raw JSON value is a string.
func isStringRaw(raw json.RawMessage) bool {
	t := bytes.TrimSpace(raw)
	return len(t) > 0 && t[0] == '"'
}

// isRawNull reports whether the raw JSON value is null.
func isRawNull(raw json.RawMessage) bool {
	return string(bytes.TrimSpace(raw)) == "null"
}

// validCancelParams implements the strict CANCEL schema: params must
// be an object with exactly one member `request_id` whose value is a
// string or an integral JSON number. The transport ignores an invalid
// cancel notification (it is a notification: no cancellation, no
// response).
func validCancelParams(params json.RawMessage) bool {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(params, &obj); err != nil || len(obj) != 1 {
		return false
	}
	raw, ok := obj["request_id"]
	if !ok {
		return false
	}
	if isStringRaw(raw) {
		return true
	}
	return isIntegralNumberRaw(raw)
}

// canonicalIntegralText returns the canonical text of an integral JSON
// number. serde_json normalizes the literal -0 to 0, so both the
// request-id echo and the cancel correlation key must use this same
// canonical text for same-batch cancellation to be byte-parity with the
// Rust binary.
func canonicalIntegralText(raw json.RawMessage) (string, bool) {
	t := bytes.TrimSpace(raw)
	if !integralText(t) {
		return "", false
	}
	if bytes.Equal(t, []byte("-0")) {
		t = []byte("0")
	}
	return string(t), true
}

// numberCancelKey maps the `cancel.request_id` number to its
// canonical key, or returns ok==false for a non-integral value.
func numberCancelKey(raw json.RawMessage) (string, bool) {
	text, ok := canonicalIntegralText(raw)
	if !ok {
		return "", false
	}
	return "n:" + text, true
}

// isIntegralNumberRaw reports whether the raw JSON value is an
// integral number literal (a JSON number with no '.', 'e', or 'E').
func isIntegralNumberRaw(raw json.RawMessage) bool {
	t := bytes.TrimSpace(raw)
	if len(t) == 0 {
		return false
	}
	c := t[0]
	if c != '-' && (c < '0' || c > '9') {
		return false
	}
	return integralText(t)
}

// validID accepts a JSON string or an integral JSON number.
func validID(raw json.RawMessage) (*RequestId, bool) {
	t := bytes.TrimSpace(raw)
	if len(t) == 0 {
		return nil, false
	}
	if t[0] == '"' {
		var s string
		if json.Unmarshal(t, &s) != nil {
			return nil, false
		}
		id := RequestIdFromString(s)
		return &id, true
	}
	if t[0] == '-' || (t[0] >= '0' && t[0] <= '9') {
		text, ok := canonicalIntegralText(t)
		if !ok {
			return nil, false
		}
		id := RequestIdFromNumber(text)
		return &id, true
	}
	return nil, false
}

func decodeEnvelope(item json.RawMessage, batchIndex *int) (*Request, *SchemaError) {
	var obj map[string]json.RawMessage
	if err := json.Unmarshal(item, &obj); err != nil {
		return nil, InvalidRequest("request must be an object")
	}
	for key := range obj {
		switch key {
		case "jsonrpc", "id", "method", "params":
		default:
			return nil, InvalidRequest(fmt.Sprintf("unknown request member %q", boundedDiagnosticText(key)))
		}
	}
	vr, ok := obj["jsonrpc"]
	if !ok || !isStringRaw(vr) {
		return nil, InvalidRequest(`jsonrpc must be "2.0"`)
	}
	var version string
	if json.Unmarshal(vr, &version) != nil || version != "2.0" {
		return nil, InvalidRequest(`jsonrpc must be "2.0"`)
	}
	mr, ok := obj["method"]
	if !ok || !isStringRaw(mr) {
		return nil, InvalidRequest("method must be a string starting with iprange.v1.")
	}
	var method string
	if err := json.Unmarshal(mr, &method); err != nil {
		return nil, InvalidRequest("method must be a string starting with iprange.v1.")
	}
	if len(method) < len(MethodPrefix) || method[:len(MethodPrefix)] != MethodPrefix {
		return nil, &SchemaError{Code: StdMethodNotFound, Message: "method must start with iprange.v1."}
	}
	var id *RequestId
	if ir, ok := obj["id"]; ok {
		if isRawNull(ir) {
			return nil, InvalidRequest("id must be a string or integral number")
		}
		valid, accepted := validID(ir)
		if !accepted {
			return nil, InvalidRequest("id must be a string or integral number")
		}
		id = valid
	}
	params, ok := obj["params"]
	if !ok || !isObjectRaw(params) {
		return nil, &SchemaError{Code: StdInvalidParams, Message: "params must be an object"}
	}
	if id == nil && method != CancelMethod {
		return nil, InvalidRequest("notifications are not accepted except iprange.v1.cancel")
	}
	return &Request{ID: id, Method: method, Params: params, BatchIndex: batchIndex}, nil
}

// DecodeFrame decodes one physical line into a single request or an
// ordered batch. The frame must already have its LF/CRLF terminator
// stripped by the LineReader.
func DecodeFrame(line []byte) ([]*Request, *SchemaError) {
	if len(line) > InputFrameLimit {
		return nil, &SchemaError{Code: TransportFrameTooLarge, Message: "frame over input limit"}
	}
	if bytes.IndexByte(line, '\n') >= 0 || bytes.IndexByte(line, '\r') >= 0 {
		return nil, ParseError("unescaped line break inside frame")
	}
	if !utf8.Valid(line) {
		return nil, ParseError("frame is not valid UTF-8")
	}
	t := bytes.TrimSpace(line)
	if len(t) == 0 {
		return nil, ParseError("parse error: empty frame")
	}
	value := json.RawMessage(t)
	// Escaped Unicode is validated before any decode or execution:
	// serde_json refuses unpaired surrogate escapes (-32700) while Go's
	// encoding/json silently substitutes U+FFFD, which would let a
	// refused frame execute with replacement bytes (schema.rs
	// decode_frame parity).
	if herr := validateSurrogates(t); herr != nil {
		return nil, ParseError("parse error: " + herr.Error())
	}
	// The preliminary parse is a lossless syntax probe (json.Decoder
	// with UseNumber): integral numbers surface as their exact lexical
	// token, so an arbitrary-precision request id decodes without
	// float64 range loss, exactly like serde_json arbitrary_precision.
	decoder := json.NewDecoder(bytes.NewReader(t))
	decoder.UseNumber()
	var probe any
	if err := decoder.Decode(&probe); err != nil {
		return nil, ParseError("parse error: " + err.Error())
	}
	if _, err := decoder.Token(); err != io.EOF {
		return nil, ParseError("parse error: multiple JSON values")
	}
	if t[0] == '[' {
		var arr []json.RawMessage
		if err := json.Unmarshal(value, &arr); err != nil {
			return nil, ParseError("parse error: invalid array")
		}
		if len(arr) == 0 || len(arr) > BatchLimit {
			return nil, InvalidRequest(fmt.Sprintf("batch length %d outside 1..%d", len(arr), BatchLimit))
		}
		requests := make([]*Request, 0, len(arr))
		for i, item := range arr {
			idx := i
			req, serr := decodeEnvelope(item, &idx)
			if serr != nil {
				return nil, serr
			}
			requests = append(requests, req)
		}
		return requests, nil
	}
	if t[0] != '{' {
		return nil, InvalidRequest("request must be an object")
	}
	req, serr := decodeEnvelope(value, nil)
	if serr != nil {
		return nil, serr
	}
	return []*Request{req}, nil
}

// encodeResponseObject encodes one response object with the
// 65,000-byte ceiling.
func encodeResponseObject(payload any) (string, *SchemaError) {
	text, err := Marshal(payload)
	if err != nil {
		return "", &SchemaError{Code: TransportFrameTooLarge, Message: "response object encoding failed"}
	}
	if len(text) > ResponseObjectLimit {
		return "", &SchemaError{Code: TransportFrameTooLarge, Message: "response object over 65,000-byte limit"}
	}
	return string(text), nil
}

// encodeResponseFrame encodes one response frame with the
// 1,048,576-byte ceiling.
func encodeResponseFrame(payload any) (string, *SchemaError) {
	text, err := Marshal(payload)
	if err != nil {
		return "", &SchemaError{Code: TransportFrameTooLarge, Message: "response frame encoding failed"}
	}
	if len(text) > OutputFrameLimit {
		return "", &SchemaError{Code: TransportFrameTooLarge, Message: "response frame over 1,048,576-byte limit"}
	}
	return string(text), nil
}

// parseHex4 decodes one JSON \uXXXX escape body; ok=false on a
// non-hex digit.
func parseHex4(body []byte) (uint16, bool) {
	var value uint16
	for _, c := range body {
		value <<= 4
		switch {
		case c >= '0' && c <= '9':
			value |= uint16(c - '0')
		case c >= 'a' && c <= 'f':
			value |= uint16(c-'a') + 10
		case c >= 'A' && c <= 'F':
			value |= uint16(c-'A') + 10
		default:
			return 0, false
		}
	}
	return value, true
}

// validateSurrogates refuses unpaired surrogate code points encoded as
// \uXXXX escapes anywhere in a JSON text, matching serde_json: a high
// surrogate (D800-DBFF) must be immediately followed by a low-surrogate
// escape (DC00-DFFF), and a lone low surrogate is refused.  Only the
// exact serde_json message families are produced ("unexpected end of
// hex escape", "lone leading surrogate in hex escape"); malformed or
// truncated escapes are left to the JSON decoder, which also refuses
// them with a parse error.
func validateSurrogates(text []byte) error {
	for i := 0; i < len(text); {
		if text[i] != '"' {
			i++
			continue
		}
		i++
		for i < len(text) {
			c := text[i]
			if c == '\\' {
				if i+1 >= len(text) {
					return nil // truncated escape: the JSON decoder refuses it
				}
				if text[i+1] != 'u' {
					i += 2
					continue
				}
				if i+6 > len(text) {
					return nil // truncated \u escape: the JSON decoder refuses it
				}
				code, ok := parseHex4(text[i+2 : i+6])
				if !ok {
					return nil // malformed escape: the JSON decoder refuses it
				}
				switch {
				case code >= 0xD800 && code <= 0xDBFF:
					rest := text[i+6:]
					if len(rest) < 6 || rest[0] != '\\' || rest[1] != 'u' {
						return errors.New("unexpected end of hex escape")
					}
					low, ok := parseHex4(rest[2:6])
					if !ok || low < 0xDC00 || low > 0xDFFF {
						return errors.New("lone leading surrogate in hex escape")
					}
					i += 12
				case code >= 0xDC00 && code <= 0xDFFF:
					return errors.New("lone leading surrogate in hex escape")
				default:
					i += 6
				}
				continue
			}
			if c == '"' {
				i++
				break
			}
			i++
		}
	}
	return nil
}

// boundedErrorResponse serializes one early schema error under both
// response bounds: the 65,000-byte object bound and, since the object
// is the whole frame, the 1,048,576-byte frame bound. A payload whose
// request-derived diagnostic cannot fit is replaced by a fixed bounded
// error: the transport never emits an unbounded request echo (Rust
// session.rs bounded schema-error path).
func boundedErrorResponse(serr *SchemaError) string {
	payload := serr.Response(nil)
	if text, err := encodeResponseObject(payload); err == nil {
		if _, ferr := encodeResponseFrame(json.RawMessage(text)); ferr == nil {
			return text
		}
	}
	fallback := (&SchemaError{
		Code:    TransportFrameTooLarge,
		Message: "error response exceeds the 65,000-byte response object limit",
	}).Response(nil)
	text, _ := encodeResponseObject(fallback)
	return text
}

// SuccessResponse builds the standard success envelope.
func SuccessResponse(id *RequestId, result any) json.RawMessage {
	m := map[string]any{
		"jsonrpc": "2.0",
		"id":      json.RawMessage(id.AsJSON()),
		"result":  result,
	}
	return mustMarshal(m)
}

// ErrorResponse builds the standard error envelope with optional data.
func ErrorResponse(id *RequestId, code int64, message string, data any) json.RawMessage {
	m := map[string]any{
		"jsonrpc": "2.0",
		"id":      json.RawMessage(id.AsJSON()),
		"error":   map[string]any{"code": code, "message": message},
	}
	if data != nil {
		m["error"].(map[string]any)["data"] = data
	}
	return mustMarshal(m)
}

// mustMarshal serializes a response envelope; the payloads here are
// built from bounded constants and validated by the callers.
func mustMarshal(v any) json.RawMessage {
	b, err := Marshal(v)
	if err != nil {
		panic("rpc: marshal of response envelope failed: " + err.Error())
	}
	return b
}

// Canonical error data payloads used by the session transport.
func outputLimitErrorData() map[string]any {
	return map[string]any{"code": "output_limit", "outcome": "read_only_failure"}
}

// EncodeResponseObjectProbe exports the object-ceiling check for the
// handler preflight helpers (envelope probes sized before any work).
func EncodeResponseObjectProbe(payload any) (string, *SchemaError) {
	return encodeResponseObject(payload)
}
